plugins/tomcat8/src/main/java/org/radargun/service/TomcatServerService.java

```diff
@@ -0,0 +1,284 @@
+      cmd.add(catalinaBase + "/conf/radargun-tomcat-" + ServiceHelper.getSlaveIndex() + ".xml");
+      cmd.add("start");
+
+      return cmd;
+   }
+
+   @Override
+   public Map<String, String> getEnvironment() {
+      Map<String, String> envs = new HashMap(super.getEnvironment());
+      if (java != null) {
+         if (envs.containsKey(JAVA_HOME)) {
+            log.warn("Overriding " + JAVA_HOME + ": " + envs.get(JAVA_HOME) + " with " + java);
+         }
+         envs.put(JAVA_HOME, java);
+      }
+      if (javaOpts != null) {
+         if (envs.containsKey(JAVA_OPTS)) {
+            log.warn("Overriding " + JAVA_OPTS + ": " + envs.get(JAVA_OPTS) + " with " + javaOpts);
+         }
+         envs.put(JAVA_OPTS, javaOpts);
+      }
+      return envs;
+   }
+
+   boolean isTomcatReady() {
+      try {
+         queryTomcat("/text/list");
+         return true;
+      } catch (final IOException e) {
+         return false;
+      }
+   }
+
+   /**
+    * Copied from Arquillian and modified.
+    *
+    * Execute the specified command, based on the configured properties. The input stream will be closed upon completion of
+    * this task, whether it was executed successfully or not.
+    *
+    * @param command Command to be executed
+    * @throws IOException
+    */
+   private void queryTomcat(final String command) throws IOException {
+      URLConnection conn = new URL(getManagerUrl() + command).openConnection();
+      final HttpURLConnection hconn = (HttpURLConnection) conn;
+
+      // Set up standard connection characteristics
+      hconn.setAllowUserInteraction(false);
+      hconn.setDoInput(true);
+      hconn.setUseCaches(false);
+      hconn.setDoOutput(false);
+      hconn.setRequestMethod("GET");
+
+      if (user != null && user.length() != 0) {
+         hconn.setRequestProperty("Authorization", constructHttpBasicAuthHeader());
+      }
+      hconn.setRequestProperty("Accept", "text/plain");
+
+      // Establish the connection with the server
+      hconn.connect();
+
+      // Send the request data (if any)
+      processResponse(command, hconn);
+   }
+
+   /**
+    * Copied from Arquillian and modified.
+    */
+   private void processResponse(final String command, final HttpURLConnection hconn) throws IOException {
+      final int httpResponseCode = hconn.getResponseCode();
+      if (httpResponseCode >= 400 && httpResponseCode < 500) {
+         throw new IllegalStateException(
+            "Unable to connect to Tomcat manager. "
+               + "The server command (" + command + ") failed with responseCode ("
+               + httpResponseCode + ") and responseMessage (" + hconn.getResponseMessage()+ ").\n\n"
+               + "Please make sure that you provided correct credentials for a user which is allowed to access Tomcat manager application.\n"
+               + "The user must have 'manager-script' role specified in tomcat-users.xml file.\n");
+      } else if (httpResponseCode >= 300) {
```

Comment:
We don't want to deal with redirections because there are many reasons why redirection happens and this code is not supposed to deal with that. Hence I consider this an error. And I believe Arqullian engineers had a similar opinion. This code is taken from there and slightly modified.